imports/wasi_snapshot_preview1/fs.go

```diff
@@ -643,11 +643,243 @@ func fdRead_shouldContinueRead(n, l uint32, err error) (bool, Errno) {
+const direntSize = uint32(24)
+
+// maxDirents returns the maximum count and total entries that can fit in
+// maxLen bytes.
+//
+// truncatedEntryLen is the amount of bytes past bufLen needed to write the
+// next entry. We have to return bufused == bufLen unless the directory is
+// exhausted.
+//
+// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#fd_readdir
+// See https://github.com/WebAssembly/wasi-libc/blob/659ff414560721b1660a19685110e484a081c3d4/libc-bottom-half/cloudlibc/src/libc/dirent/readdir.c#L44
+func maxDirents(entries []fs.DirEntry, bufLen uint32) (bufused, direntCount, truncatedEntryLen uint32) {
+	lenRemaining := bufLen
+	for _, e := range entries {
+		nameLen := uint32(len(e.Name()))
+		entryLen := direntSize + nameLen
+
+		if lenRemaining == 0 {
+			// We've hit exactly the end of bufLen, so cannot write the next
+			// entry. A caller who wants more will retry.
+			break
+		}
+
+		if lenRemaining < entryLen {
+			// We haven't room to write the entry, and docs say to write the
+			// header. This helps especially when there is an entry with a very
+			// long filename. Ex if bufLen is 4096 and the filename is 4096,
+			// we need to write direntSize(24) + 4096 bytes to write the entry.
+			// In this case, we only write up to direntSize(24) to allow the
+			// caller to resize.
+
+			// bufused != bufLen means more entries exist, which is the case
+			// when the next entry is larger than bytes remaining.
+			bufused = bufLen
+
+			// When the next entry is truncated, only write the dirent.
+			if lenRemaining < direntSize {
+				truncatedEntryLen = direntSize - lenRemaining
+			}
+			break
+		}
+
+		// At this point, this entry can fit within bufLen.
+		bufused += entryLen
+		lenRemaining -= entryLen
+		direntCount++
+	}
+	return
+}
+
+// writeDirents writes the directory entries to the buffer, which is pre-sized
+// based on maxDirents.	truncatedEntryLen means write one past entryCount,
+// without its name. See maxDirents for why
+func writeDirents(
+	entries []fs.DirEntry,
+	entryCount uint32,
+	truncatedEntryLen uint32,
+	dirents []byte,
+	cookie uint64,
+) {
+	pos, i := uint32(0), uint32(0)
+	for ; i < entryCount; i++ {
+		e := entries[i]
+		nameLen := uint32(len(e.Name()))
+
+		writeDirent(dirents[pos:], cookie, nameLen, e.IsDir())
+		pos += direntSize
+
+		copy(dirents[pos:], e.Name())
+		pos += nameLen
+		cookie++
+	}
+
+	if truncatedEntryLen == 0 {
+		return
+	}
+
+	// Write a dirent without its name
+	dirent := make([]byte, direntSize)
+	e := entries[i]
+	writeDirent(dirent, cookie, uint32(len(e.Name())), e.IsDir())
+
+	// Potentially truncate it
+	copy(dirents[pos:], dirent)
+}
+
+// writeDirent writes direntSize bytes
+func writeDirent(buf []byte, dNext uint64, dNamlen uint32, dType bool) {
+	binary.LittleEndian.PutUint64(buf, dNext)        // d_next
+	binary.LittleEndian.PutUint64(buf[8:], 0)        // no d_ino
+	binary.LittleEndian.PutUint32(buf[16:], dNamlen) // d_namlen
```

Comment:
calling PutUintXX three times might be costly, but instead how about directly putting integers into buf as we know that it is little endian